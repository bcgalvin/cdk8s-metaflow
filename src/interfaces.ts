import { HttpIngressPathType, ImagePullPolicy, ServiceType } from './enums';
import { Affinity, EnvFromSource, IngressTls, SecurityContext, Toleration } from './imports/k8s';

export interface ImageOptions {
  readonly repository?: string;
  readonly pullPolicy?: ImagePullPolicy;
  readonly tag?: string;
}

export interface ServiceAccountOptions {
  readonly create?: boolean | undefined;
  readonly annotations?: Record<string, string>;
  readonly name?: string;
}

export interface DbMigrationOptions {
  readonly runOnStart?: boolean;
  readonly onlyIfDbEmpty?: boolean;
}

export interface ServiceOptions {
  readonly type?: ServiceType;
  readonly port?: number;
  readonly annotations?: Record<string, string>;
}

export interface HttpIngressPath {
  readonly path?: string;
  readonly pathType?: HttpIngressPathType;
}

export interface HostOptions {
  readonly host?: string;
  readonly paths?: HttpIngressPath[];
}

export interface IngressOptions {
  readonly enabled?: boolean;
  readonly className?: string;
  readonly annotations?: Record<string, string>;
  readonly hosts?: HostOptions[];
  readonly tls?: IngressTls;
}

export interface AutoscalingOptions {
  readonly enabled?: boolean;
  readonly minReplicas?: number;
  readonly maxReplicas?: number;
  readonly targetCPUUtilizationPercentage?: number;
}

export interface IApiResource {
  readonly apiGroup?: string;
  readonly resourceType?: string;
  readonly resourceName?: string;
}

export interface MetadataDatabaseEnvOptions {
  readonly port?: number;
  readonly password?: string;
  readonly host?: string;
  readonly user?: string;
}

export interface MetaflowServiceOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly dbMigrations?: DbMigrationOptions;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
}

export interface MetaflowUIOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly uiImage?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly env: Record<string, string>;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly serviceStatic?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
  readonly metaflowDatastoreSysrootS3?: string;
}

export interface DatabaseAuthOptions {
  readonly database?: string;
  readonly password?: string;
  readonly username?: string;
  readonly postgresPassword?: string;
  readonly enablePostgresUser?: boolean;
  readonly replicationPassword?: string;
  readonly replicationUsername?: string;
}

export interface DatabaseMetricsOptions {
  readonly enabled?: boolean;
}

export interface DatabaseResourceRequestOptions {
  readonly memory?: string;
  readonly cpu?: string;
}

export interface DatabaseResourcesOptions {
  readonly requests?: DatabaseResourceRequestOptions;
}

export interface DatabaseReplicationOptions {
  readonly enabled?: boolean;
  readonly readReplicas?: number;
}

export interface DatabaseVolumePermissionsOptions {
  readonly enabled?: boolean;
}

export interface MetadataDatabaseOptions {
  readonly auth?: DatabaseAuthOptions;
  readonly metrics?: DatabaseMetricsOptions;
  readonly resources?: DatabaseResourcesOptions;
  readonly replication?: DatabaseReplicationOptions;
  readonly architecture?: string;
  readonly volumePermissions?: DatabaseVolumePermissionsOptions;
}
