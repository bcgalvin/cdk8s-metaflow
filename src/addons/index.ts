export * from './addon';
export * from './postgres';
export * from './argo-workflows';
export * from './autoscaler';
