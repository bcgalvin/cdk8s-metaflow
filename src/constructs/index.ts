export * from './metaflow-service';
export * from './interfaces';
