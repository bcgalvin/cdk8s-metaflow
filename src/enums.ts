export enum ImagePullPolicy {
  ALWAYS = 'Always',
  IF_NOT_PRESENT = 'IfNotPresent',
  NEVER = 'Never',
}

export enum HttpIngressPathType {
  PREFIX = 'Prefix',
  EXACT = 'Exact',
  IMPLEMENTATION_SPECIFIC = 'ImplementationSpecific',
}

export enum ServiceType {
  CLUSTER_IP = 'ClusterIP',
  NODE_PORT = 'NodePort',
  LOAD_BALANCER = 'LoadBalancer',
  EXTERNAL_NAME = 'ExternalName',
}
