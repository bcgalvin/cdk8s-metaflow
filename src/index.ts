export * from './constructs';
export * from './addons';
