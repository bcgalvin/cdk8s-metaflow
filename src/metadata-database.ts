import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetadataDatabaseOptions } from './interfaces';

export interface MetadataDatabaseProps {
  readonly chartVersion: string;
  readonly chartValues?: MetadataDatabaseOptions;
}

export class MetadataDatabase extends Construct {
  constructor(scope: Construct, id: string, props: MetadataDatabaseProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {
      enabled: true,
      auth: {
        database: 'metaflow',
        password: 'metaflow',
        username: 'metaflow',
      },
    };

    new Helm(this, 'postgres', {
      chart: 'bitnami/postgresql',
      releaseName: 'release-name-postgresql',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
