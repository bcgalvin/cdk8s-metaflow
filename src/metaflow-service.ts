import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetaflowServiceOptions } from './interfaces';

export interface MetaflowServiceProps {
  readonly chartVersion: string;
  readonly chartValues?: MetaflowServiceOptions;
}

export class MetaflowService extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowServiceProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {};

    new Helm(this, 'metaflow-service-helm', {
      chart: 'metaflow/metaflow-service',
      releaseName: 'release-name-metaflow-service',
      helmFlags: ['--version', props.chartVersion],
      values: { ...chartValues },
    });
  }
}
