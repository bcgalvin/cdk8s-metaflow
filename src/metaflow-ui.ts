import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetaflowUIOptions } from './interfaces';

export interface MetaflowUIProps {
  readonly chartVersion: string;
  readonly chartValues?: MetaflowUIOptions;
}

export class MetaflowUI extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowUIProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {};

    new Helm(this, 'metaflow-ui-helm', {
      chart: 'metaflow/metaflow-ui',
      releaseName: 'release-name-metaflow-ui',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
