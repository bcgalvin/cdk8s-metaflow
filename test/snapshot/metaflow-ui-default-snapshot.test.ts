import { Chart, Testing } from 'cdk8s';
import { ServiceAccount, ServiceType } from 'cdk8s-plus-22';
import { MetaflowUIChart } from '../../src/';

describe('metaflow ui', () => {
  test('default snapshot', () => {
    const app = Testing.app();
    const baseChart = new Chart(app, 'base');
    const sa = new ServiceAccount(baseChart, 'metaflow-ui-sa', {});

    // GIVEN
    const chart = new MetaflowUIChart(app, 'test-service', {
      serviceAccount: sa,
      image: 'netflixoss/metaflow_metadata_service',
      imageTag: 'v2.3.0',
      serviceType: ServiceType.CLUSTER_IP,
    });
    // THEN
    expect(chart).toMatchSnapshot();
  });
});
