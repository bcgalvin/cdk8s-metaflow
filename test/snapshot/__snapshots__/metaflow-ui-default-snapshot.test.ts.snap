// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaflow ui default snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 8083,
          "protocol": "TCP",
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/name": "metaflow-ui",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui-static",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui-static",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 3000,
          "protocol": "TCP",
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/name": "metaflow-ui-static",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "release-name-metaflow-ui",
          "app.kubernetes.io/name": "metaflow-ui",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "release-name-metaflow-ui",
            "app.kubernetes.io/name": "metaflow-ui",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "/opt/latest/bin/python3",
                "-m",
                "services.ui_backend_service.ui_server",
              ],
              "env": Array [
                Object {
                  "name": "UI_ENABLED",
                  "value": "1",
                },
                Object {
                  "name": "PATH_PREFIX",
                  "value": "/api",
                },
                Object {
                  "name": "MF_DATASTORE_ROOT",
                  "value": "",
                },
                Object {
                  "name": "METAFLOW_DATASTORE_SYSROOT_S3",
                  "value": "",
                },
                Object {
                  "name": "LOGLEVEL",
                  "value": "DEBUG",
                },
                Object {
                  "name": "METAFLOW_SERVICE_URL",
                  "value": "http://localhost:8083/api/metadata",
                },
                Object {
                  "name": "METAFLOW_DEFAULT_DATASTORE",
                  "value": "s3",
                },
                Object {
                  "name": "METAFLOW_DEFAULT_METADATA",
                  "value": "service",
                },
                Object {
                  "name": "MF_METADATA_DB_NAME",
                },
                Object {
                  "name": "MF_METADATA_DB_PORT",
                  "value": "5432",
                },
                Object {
                  "name": "MF_METADATA_DB_PSWD",
                  "value": "",
                },
                Object {
                  "name": "MF_METADATA_DB_USER",
                  "value": "",
                },
                Object {
                  "name": "MF_METADATA_DB_HOST",
                  "value": "release-name-metaflow-ui-postgresql",
                },
              ],
              "image": "netflixoss/metaflow_metadata_service:2.1.0",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "/api/ping",
                  "port": "http",
                },
              },
              "name": "metaflow-ui",
              "ports": Array [
                Object {
                  "containerPort": 8083,
                  "name": "http",
                  "protocol": "TCP",
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "/api/ping",
                  "port": "http",
                },
              },
              "resources": Object {},
              "securityContext": Object {},
            },
          ],
          "securityContext": Object {},
          "serviceAccountName": "release-name-metaflow-ui",
        },
      },
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui-static",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui-static",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "release-name-metaflow-ui",
          "app.kubernetes.io/name": "metaflow-ui-static",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "release-name-metaflow-ui",
            "app.kubernetes.io/name": "metaflow-ui-static",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "public.ecr.aws/outerbounds/metaflow_ui:v1.0.1",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "name": "metaflow-ui-static",
              "ports": Array [
                Object {
                  "containerPort": 3000,
                  "name": "http",
                  "protocol": "TCP",
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "/",
                  "port": "http",
                },
              },
              "resources": Object {},
              "securityContext": Object {},
            },
          ],
          "securityContext": Object {},
          "serviceAccountName": "release-name-metaflow-ui",
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "annotations": Object {
        "helm.sh/hook": "test",
      },
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-metaflow-ui",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-ui",
        "app.kubernetes.io/version": "2.1.0",
        "helm.sh/chart": "metaflow-ui-0.1.2",
      },
      "name": "release-name-metaflow-ui-test-connection",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "args": Array [
            "release-name-metaflow-ui:8083",
          ],
          "command": Array [
            "wget",
          ],
          "image": "busybox",
          "name": "wget",
        },
      ],
      "restartPolicy": "Never",
    },
  },
]
`;
